grpc/src/main/java/io/stargate/grpc/idempotency/IdempotencyAnalyzer.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.grpc.idempotency;
+
+import io.stargate.proto.QueryOuterClass;
+import java.util.Arrays;
+import java.util.List;
+
+public class IdempotencyAnalyzer {
```

Comment:
We may consider moving this to the `persistence-api` and propagating the `idempotency` info in the:
`public static class Prepared extends Result`
The `isIdempotent` analysis logic seems not gRPC specific, so other modules that are preparing using cql String may benefit from that. wdyt?

Reply:
After some more analysis, I found that it may be hard to move the `isIdempotent` to the `Result.Prepared` class. 
The `toResultInternal` method is constructing the `Preapred` in the 
```
public static Result toResultInternal(
      ResultMessage resultMessage, org.apache.cassandra.transport.ProtocolVersion version)
``` 
method. I don't see a way to extract the query from the `resultMessage` to analyze whether it is idempotent or not.
We could propagate the `String query` up from:
1. `public CompletableFuture<Result.Prepared> prepare(String query, Parameters parameters)`
2. `private <T extends Result> CompletableFuture<T> executeRequestOnExecutor(
            Parameters parameters, long queryStartNanoTime, Supplier<Request> requestSupplier,
            @Nullable String query)`
3. ```
 Conversion.toResult(
                        (ResultMessage) response,
                        Conversion.toInternal(parameters.protocolVersion()),
                            query);```
4. ```
public static Result toResultInternal(
          ResultMessage resultMessage, org.apache.cassandra.transport.ProtocolVersion version,
          String query)
```
5. and finally:
```
return new Result.Prepared(
            Conversion.toExternal(prepared.statementId),
            null,
            toResultMetadata(prepared.resultMetadata, null),
            toPreparedMetadata(prepared.metadata.names, preparedStatement.partitionKeyBindIndexes),
                IdempotencyAnalyzer.isIdempotent(query)
                );
```
but it seems that it complicates the code of persistence too much, outweighing benefits.

The other solution would be to propagate the `Request` and it to `PrependMessage`:
```
PrepareMessage prepareMessage = (PrepareMessage) request;
```
as the `PrepareMessage` has a `query` object underneath. Unfortunately, it does not expose it in any way, so we would need to use the reflection to extract the `query` object. It  I am not sure if this use case justifies it.